### PR TITLE
[SPARK-35921][BUILD] ${spark.yarn.isHadoopProvided} in config.properties is not edited if build with SBT

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR changes `SparkBuild.scala` to edit `config.properties` in `yarn` sub-module in build with SBT like as build with Maven does.

### Why are the changes needed?

yarn sub-module contains config.properties.
```
spark.yarn.isHadoopProvided = ${spark.yarn.isHadoopProvided}
```

The `${spark.yarn.isHadoopProvided}` part is replaced with `true` or `false` in build depending on whether Hadoop is provided or not (specified by -Phadoop-provided).
The edited config.properties will be loaded at runtime to control how to populate Hadoop-related classpath.

If we build with Maven, these process works but doesn't with SBT.

If we build with SBT and deploy apps on YARN, the following warning appears and classpath is not populated correctly.
```
21/06/29 10:51:20 WARN config.package: Can not load the default value of `spark.yarn.isHadoopProvided` from `org/apache/spark/deploy/yarn/config.properties` with error, java.lang.IllegalArgumentException: For input string: "${spark.yarn.isHadoopProvided}". Using `false` as a default value.
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Built with SBT and extracted `config.properties` from the build artifact and confirmed `${spark.yarn.isHadoopProvided} was correctly edited with `true` or `false`.
```
cat org/apache/spark/deploy/yarn/config.properties 
spark.yarn.isHadoopProvided = false                                # In case build with -Pyarn and without -Phadoop-provided
spark.yarn.isHadoopProvided = true                                 # In case build with -Pyarn and -Phadoop-provided
```
I also confirmed the warning message shown above no longer appears.

### CODE REVIEW

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -802,11 +805,29 @@ object Hive {
...
+    "Generate config.properties which contains a setting whether Hadoop is provided or not")
+  val propFileName = "config.properties"
+
   lazy val settings = Seq(
     excludeDependencies --= Seq(
       ExclusionRule(organization = "com.sun.jersey"),
       ExclusionRule("javax.servlet", "javax.servlet-api"),
-      ExclusionRule("javax.ws.rs", "jsr311-api"))
+      ExclusionRule("javax.ws.rs", "jsr311-api")),
+      Compile / unmanagedResources :=
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala:817]: nit but looks like should indent 2 spaces back for these

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -413,6 +413,9 @@ object SparkBuild extends PomBuild {
 
   enable(KubernetesIntegrationTests.settings)(kubernetesIntegrationTests)
 
+  if (profiles.contains("hadoop-provided")) {
+    sys.props("spark.yarn.isHadoopProvided") = "true"
+  }
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala:418]: nit but would it be better to move this logic into the `YARN`?

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -802,11 +805,29 @@ object Hive {
...
+      Compile / unmanagedResources :=
+        (Compile / unmanagedResources).value.filter(!_.getName.endsWith(s"$propFileName")),
+      genConfigProperties := {
+        val file = (Compile / classDirectory).value / s"org/apache/spark/deploy/yarn/$propFileName"
+        val isHadoopProvided = sys.props.getOrElse("spark.yarn.isHadoopProvided", "false")
+        IO.write(file, s"spark.yarn.isHadoopProvided = $isHadoopProvided")
+      },
+      (Compile / copyResources) := (Def.taskDyn {
+        val c = (Compile / copyResources).value
+          Def.task {
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala:826]: looks like we should indent 2 spaces back here too

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
cc @dbtsai FYI. It's more a followup of https://github.com/apache/spark/pull/28788

File: project/SparkBuild.scala (hunk 2 above)
Comment (sarutak) [reply project/SparkBuild.scala]: `profiles` cannot be referred from `YARN` but I found another better way.

File: project/SparkBuild.scala (hunk 3 above)
Comment (sarutak) [reply project/SparkBuild.scala]: Oh, I was confused because `Seq` is nested... Thanks.

File: project/SparkBuild.scala (hunk 1 above)
Comment (sarutak) [reply project/SparkBuild.scala]: Oh, I was confused because `Seq` is nested... Thanks.

File: project/SparkBuild.scala [hunk 4]
```diff
@@ -802,11 +802,30 @@ object Hive {
...
+    Compile / unmanagedResources :=
+      (Compile / unmanagedResources).value.filter(!_.getName.endsWith(s"$propFileName")),
+    genConfigProperties := {
+      val file = (Compile / classDirectory).value / s"org/apache/spark/deploy/yarn/$propFileName"
+      val isHadoopProvided = SbtPomKeys.effectivePom.value.getProperties.get(hadoopProvidedProp)
+      IO.write(file, s"$hadoopProvidedProp = $isHadoopProvided")
+    },
+    (Compile / copyResources) := (Def.taskDyn {
+      val c = (Compile / copyResources).value
+        Def.task {
```
Comment (sarutak) [inline project/SparkBuild.scala:824]: Ah, still mis-intended. I'll fix it.

**Comment by HyukjinKwon** [comment]
LGTM

**Comment by dbtsai** [comment]
LGTM too.

**Comment by dbtsai** [comment]
Merged into master for Spark 3.2 release. Thanks.

**Pull request closed without merging**